app/controllers/portal/tax_returns_controller.rb

```diff
@@ -0,0 +1,72 @@
+module Portal
+  class TaxReturnsController < PortalController
+    before_action :load_client_tax_return, except: [:success]
+    before_action :redirect_unless_spouse_signature_required, only: [:spouse_sign, :spouse_authorize_signature]
+    before_action :redirect_unless_primary_signature_required, only: [:sign, :authorize_signature]
+    layout "application"
+
+    def authorize_signature
+      @form = Portal::PrimarySignForm8879.new(@tax_return)
+    end
+
+    def spouse_authorize_signature
+      @form = Portal::SpouseSignForm8879.new(@tax_return)
+    end
+
+    def sign
+      form_class = Portal::PrimarySignForm8879
+      @form = form_class.new(@tax_return, permitted_params(form_class))
+      if @form.sign
+        flash[:success] = I18n.t("controllers.tax_returns_controller.success", year: @tax_return.year)
+        redirect_to :portal_root
+      else
+        flash.now[:alert] = I18n.t("controllers.tax_returns_controller.errors.#{@form.errors.keys.first}")
+        render :authorize_signature
+      end
+    end
+
+    def spouse_sign
+      form_class = Portal::SpouseSignForm8879
+      @form = form_class.new(@tax_return, permitted_params(form_class))
+      if @form.sign
+        flash[:success] = I18n.t("controllers.tax_returns_controller.success", year: @tax_return.year)
+        redirect_to :portal_root
+      else
+        flash.now[:alert] = I18n.t("controllers.tax_returns_controller.errors.#{@form.errors.keys.first}")
+        render :spouse_authorize_signature
+      end
+    end
+
+    def success; end
+
+    private
+
+    def load_client_tax_return
+      @tax_return = current_client.tax_returns.includes(client: [:intake]).find(params[:tax_return_id])
```

Comment:
I like getting nil back rather than exceptions if possible. Purely aesthetic I'd say and just my personal preference.